datadogfirehosenozzle/datadog_firehose_nozzle.go

```diff
@@ -103,16 +104,35 @@ func (d *DatadogFirehoseNozzle) Start() error {
 	d.cfClient = d.createCFClient()
 	// Create metric processor instance
 	d.processor = d.createProcessor()
-	// Start consumer from the Firehose
-	err = d.consumeFirehose(authToken)
-	if err != nil {
-		return err
+
+	// Let's store the last three times when a restart occured, so that we don't keep trying forever
+	retryHistory = make([]time.Time, 3)
+	for {
+		// Start consumer from the Firehose
+		err = d.consumeFirehose(authToken)
+		if err != nil {
+			return err
+		}
+		// Start multiple workers to parallelize firehose events (event.envelope) transformation into processedMetrics
+		// and then grouped into metricsMap
+		d.startWorkers()
+		// Start infinite loop to periodically submit metrics (metricsMap) to Datadog
+		err = d.postToDatadog()
+		if !shouldReconnect(err) {
+			break
+		}
+
+		// memorize the retry timestamp
+		retryHistory = append(retryHistory[1:], time.Now())
+		if retryHistory[2].Sub(retryHistory[0]) < 5*time.Second {
```

Comment:
what if retryHistory[2] doesn't exist? Will this error?

Reply:
I initialize the slice with 3 elements, and the logic makes it that there are always 3 elements